transmart-rest-api/grails-app/controllers/org/transmartproject/rest/UserQuerySetController.groovy

```diff
@@ -0,0 +1,107 @@
+package org.transmartproject.rest
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.web.bind.annotation.PathVariable
+import org.transmartproject.core.userquery.UserQuerySetDiff
+import org.transmartproject.core.userquery.UserQuerySetInstance
+import org.transmartproject.core.userquery.UserQuerySetResource
+import org.transmartproject.rest.misc.CurrentUser
+
+import static org.transmartproject.rest.misc.RequestUtils.checkForUnsupportedParams
+
+class UserQuerySetController {
+
+    @Autowired
+    CurrentUser currentUser
+
+    @Autowired
+    UserQuerySetResource userQuerySetResource
+
+    static responseFormats = ['json']
+
+    /**
+     * Scans for changes in results of the stored queries and updates stored sets:
+     * <code>/${apiVersion}/query_sets/scan</code>
+     *
+     * This endpoint should be called after loading, deleting or updating the data.
+     * Only available for administrators.
+     *
+     * @return number of sets that were updated, which is also a number of created querySets
+     */
+    def scan() {
+        Integer result = userQuerySetResource.scan(currentUser)
+        response.status = 201
+        respond([numberOfUpdatedSets: result])
+    }
+
+    /**
+     * Gets a list of query sets related to a specific query:
+     * <code>/${apiVersion}/query_sets/${queryId}</code>
+     *
+     * @param firstResult - first result - parameter required for pagination, by default equals 0
+     * @param numResults - number of results - parameter required for pagination
+     * @param queryId - id of the query
+     * @return list of querySets
+     */
+    def getSetsByQueryId(@PathVariable('queryId') Long queryId) {
+        checkForUnsupportedParams(params, ['queryId', 'firstResult', 'numResults'])
+        int firstResult = params.firstResult ?: 0
+        def numResults = params.numResults as Integer
+
+        List<UserQuerySetInstance> querySetInstances = userQuerySetResource.getSetInstancesByQueryId(queryId, currentUser,
+                firstResult, numResults)
+        def querySetInstancesMap = querySetInstances?.groupBy { it.querySet }
+        respond([querySetInstances: querySetInstancesMap ?
+                querySetInstancesMap.collect { toResponseMap(it, UserQuerySetInstance.class) } : []])
+    }
+
+    /**
+     * Gets a list of query result change entries by query id - history of data sets changes for specific query:
+     * <code>/${apiVersion}/query_sets/diffs/${queryId}</code>
+     *
+     * @param firstResult - first result - parameter required for pagination, by default equals 0
+     * @param numResults - number of results - parameter required for pagination
+     * @param queryId - id of the query
+     * @return list of queryDiffs
+     */
```

Comment:
What is paginated here, the object ids, or the snapshots/diffs?

Reply:
These object ids are quite meaningless numbers (patient nums are generated by the database, automatically incremented).
- Two questions: why would you want to expose those in the first place?
- Do you want to send them every time you send the list of diffs, or would it be better to have a separate call to retrieve the patients that are added/removed in a specific diff, e.g, `/v2/queries/${queryId}/diffs/${diffId}/changes`.